### ISSUE TITLE
How to construct a toeplitz matrix inside a jit-able function?

### ISSUE THREAD

**Issue by golobor** [google/jax#1646]
Hi,
many thanks for this wonderful project!
Is there any possible way to construct a toeplitz matrix inside a jit-able differentiable function? 

My goal is to fit a large matrix with a product of a toeplitz matrix with a rank 1 matrix. I was hoping to achieve that by using jax autograd functionality to calculate the loss function and its gradient+hessian. The only obstacle is building a toeplitz matrix inside the loss function. There does not seem any support for toeplitz matrices in jax, so, instead, I tried iterating over diagonals of the matrix in a loop. However, jit-ing such function takes several minutes, presumably b/c jit explicitly unrolls these loops. Is there any trick to build a toeplitz matrix using jax?

Thank you,
Anton.

**Comment by mattjj** [comment]
Great question! (And thanks for the kind words!)

Just to make sure I understand: you tried writing a loop that to make an NxN Toeplitz matrix required 2N-1 steps (one for each diagonal), and that was too slow to compile? If so, I agree that it must be the unrolled loop.

The direct thing to do might be to write the same loop in terms of the slightly-more-awkward [`lax.fori_loop`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.fori_loop.html#jax.lax.fori_loop).

I wonder if there's some clever trick we can use involving convolution...

**Comment by mattjj** [comment]
About that convolution idea... here are some things that may be true:

1. a toeplitz matrix can be embedded in a larger circulant matrix (with some zeros), and so if we could construct circulant matrices efficiently maybe we could extract a block to get the toeplitz matrix we wanted;
1. the matrix for the linear map `x -> x conv y` is a circulant matrix (for circular convolution `conv`);
1. we can get the matrix of a linear map (indeed the Jacobian of any map) efficiently using `jvp`/`vjp` together with `vmap`, and particular getting the Jacobian of a conv should itself require just a single call to conv (i.e. a single primitive kernel call)

If those are true, they seem to be an argument that we can construct a circulant matrix in a single call to `conv` (with some batch dimensions), and then grab a block to make a toeplitz matrix.

Maybe there's a more direct way to think of the construction, but I wanted to guess if it was possible.

So there _could_ be something clever to do. But a loop might be good too :)

**Comment by mattjj** [comment]
Here's a much more elementary way to phrase that idea: if we conv against a one-hot vector we get a shifted version of the kernel. Given a kernel `[a b c d e]` we can use a batched version of that trick to make a matrix like

```
[[a b c d e 0 0 0]
 [0 a b c d e 0 0]
 [0 0 a b c d e 0]
 [0 0 0 a b c d e]]
```

and a submatrix is

```
[[c d e]
 [b c d]
 [a b c]]
```

**Comment by hawkinsp** [comment]
Here's another way to construct that matrix in @mattjj 's last post:
Start with:
```
[a b c d e]
```
broadcast and pad to
```
[[a b c d e 0 0 0 0]
 [a b c d e 0 0 0 0]
 ...]]
```
reshape to rank 1 (ravel) the matrix:
```
[a b c d e 0 0 0 0 a b c d e 0 0 0 0  ... ]
```
and then reshape back to rank 2 with one fewer column:
```
[[a b c d e 0 0 0]
 [0 a b c d e 0 0]
 [0 0 a b c d e 0]
 ...]]
```

**Comment by jasontuna** [comment]
Not sure if this helps, but here is a way I have used,
 
Suppose your Toeplitz matrix is mxn
and w is an (m+n-1) array with the values for the diagonals starting from the lower left to the upper right.

Then you can form the Toeplitz matrix through indexing,

```python
a =np.array((-1)*np.arange(0,m)).reshape(m,1) 
b =np.array([np.arange(m-1,m+n-1 ),])
indx = a + b
Tw=w[indx]
```

Here is an example showing jit works, multiplying a Toeplitz matrix with a vector.  

```python
import jax.random as jr
from jax import jit
import jax.numpy as np
import time

@jit
def mm(w,x):
    n=x.shape[0]
    m=w.shape[0]-n+1
    a =np.array((-1)*np.arange(0,m)).reshape(m,1)
    b =np.array([np.arange(m-1,m+n-1 ),])
    indx = a + b
    Tw=w[indx]
    return np.einsum('ij,j->i', Tw, x) 

m=600
n=800
key = jr.PRNGKey(0)
x=jr.normal(key,shape=(n,))
w=jr.normal(key,shape=(m+n-1,))
start_time = time.time()
Tw=mm(w,x)
end_time = time.time() - start_time
print("In {:0.2f} sec".format(end_time))

```

**Comment by golobor** [comment]
@jasontuna and @hawkinsp , thank you so much for putting your time into these detailed answers! 
It will take me some time to test them, but these tricks seems very promising!

**Comment by hawkinsp** [comment]
I'm going to close this issue, since it seems like there are some good solutions. Feel free to reopen if you have any more problems!

**Comment by mattjj** [comment]
@blakehechtman recently took a look at this problem, and tried these approaches:

```python
@jax.jit
def toeplitz(x):
  if len(x.shape) == 1:
    return toeplitz(x[:,None])
  n = x.shape[-2]
  m = x.shape[-1]
  if m == n:
    return x
  if m > n:
    return x[...,:n]
  r  = jnp.roll(x,m,axis=-2)
  return toeplitz(jnp.concatenate([x,r], axis=-1))

@jax.jit
def toeplitz2(x):
  n = x.shape[-1]
  iota = jnp.arange(n)
  def roll(x,i):
    return jnp.roll(x, i, axis=-1)
  return jax.vmap(roll,in_axes=(None,0),out_axes=1)(x, iota)
```

It seems the first one is best on TPU.
